Rx/v2/src/rxcpp/rx-observable.hpp

```diff
@@ -263,27 +263,107 @@ class blocking_observable
         return blocking_subscribe(source, std::forward<ArgN>(an)...);
     }
 
+    /*! Return the first item emitted by this blocking_observable, or throw an std::runtime_error exception if it emits no items.
+
+        \return  The first item emitted by this blocking_observable.
+
+        \note  If the source observable calls on_error, the raised exception is rethrown by this method.
+
+        \sample
+        When the source observable emits at least one item:
+        \snippet blocking_observable.cpp blocking first sample
+        \snippet output.txt blocking first sample
+
+        When the source observable is empty:
+        \snippet blocking_observable.cpp blocking first empty sample
+        \snippet output.txt blocking first empty sample
+    */
     T first() {
-        rxu::maybe<T> result;
-        composite_subscription cs;
-        subscribe(cs, [&](T v){result.reset(v); cs.unsubscribe();});
-        if (result.empty()) abort();
-        return result.get();
+        return source.first().as_blocking().last();
     }
 
+    /*! Return the last item emitted by this blocking_observable, or throw an std::runtime_error exception if it emits no items.
+
+        \return  The last item emitted by this blocking_observable.
+
+        \note  If the source observable calls on_error, the raised exception is rethrown by this method.
+
+        \sample
+        When the source observable emits at least one item:
+        \snippet blocking_observable.cpp blocking last sample
+        \snippet output.txt blocking last sample
+
+        When the source observable is empty:
+        \snippet blocking_observable.cpp blocking last empty sample
+        \snippet output.txt blocking last empty sample
+    */
     T last() const {
         rxu::maybe<T> result;
-        subscribe([&](T v){result.reset(v);});
-        if (result.empty()) abort();
+        rxu::maybe<std::exception_ptr> error;
+        subscribe(
+            [&](T v){result.reset(v);},
+            [&](std::exception_ptr ep){error.reset(ep);});
+        if (!error.empty())
+            std::rethrow_exception(error.get());
+        if (result.empty())
+            throw rxcpp::empty_error("No elements");
```

Comment:
I think that the string passed to empty_error should not just say the same thing as the name 'empty_error'.
I would like something like this
`rxcpp::empty_error("last() requires a stream with at least one value")`